corehq/apps/users/util.py

```diff
@@ -293,6 +293,8 @@ def update_last_sync(user, app_id, sync_date, version):
 
 
 def _last_sync_needs_update(last_sync, sync_datetime):
+    if not sync_datetime:
+        return False
```

Comment:
I'm a bit on the fence about this line. Do we ever expect `sync_datetime` to be falsy now that it's always being set to something? I'm worried this line will essentially silently swallow any future errors. For example, if this had already been here before today, we never would have gotten the errors that led you to find that heartbeat wasn't always adding the sync_date correctly, right?

If what I'm saying is on the right track, maybe it's worth reporting the error to sentry so it's still visible that something unexpected is going on.

Reply:
It's a good point, made me think more and indeed I found the bug that I introduced in https://github.com/dimagi/commcare-hq/pull/26305/. I made fixes for that here now.

Reply:
Also reverted this change, which is good to have for the reasons you have stated. Thanks!